### ISSUE TITLE
bitshift typing on unsigned operands is tedious to work with

### ISSUE THREAD

**Issue by pitrou** [numpy/numpy#8002]
```
>>> np.uint64(18446744073709551615) >> 2
Traceback (most recent call last):
  File "<ipython-input-10-988f6401a88f>", line 1, in <module>
    np.uint64(18446744073709551615) >> 2
TypeError: ufunc 'right_shift' not supported for the input types, and the inputs could not be safely coerced to any supported types according to the casting rule ''safe''
```

**Comment by pitrou** [comment]
In Numba we are considering typing based only on the first operand, see https://github.com/numba/numba/pull/2072

**Comment by sklam** [comment]
Numpy is probably casting `(unsigned, signed)` to `(float64, float64)`, which is not defined.  Then,  it tries to downcast it back to integer, but it is unsafe.

**Comment by ahaldane** [comment]
I've hit the same problem. I have thought it would have been better to copy C typing as closely as possible. Then for an algorithm already written in C I can just paste it into python, and not have to worry about wrapping all the constants in uint64.

**Comment by seberg** [comment]
Effectively fixed by NEP 50 in main/NumPy 2.0.  Closing.
